### PR TITLE
WooCommerce: Add tax settings view, including persisting to server

### LINKED ISSUE #13483: WooCommerce: Get stores settings groups into state
Fetch the merchant's stores' settings groups (general, products, tax, shipping, etc) into state so that we can eventually wire them up to containers and components
 
#### Context / Source
WooCommerce has many settings that need to be able to be adjusted by merchants using Calypso to manage their store(s). REST API endpoints already exist in WooCommerce 3.0 (and higher) just for this purpose, e.g. GET /wp-json/wc/v2/settings etc etc

### LINKED ISSUE #14439: WooCommerce: Taxes: My store is in a state that does not have a sales tax
This is the easy case. Merchants in certain states are NOT required to collect sales tax at all (i.e. Delaware, New Hampshire, Oregon).

We should probably have
- a really simple nux flow for them just to acknowledge that / remind them of that
- a super simplified no-taxes-for-you tax settings page

See also #14440 and #14441

### LINKED ISSUE #14440: WooCommerce: Taxes: My store is in a state that charges sales tax based on my business address
Merchants in certain states are required to collect sales tax based solely on the business' address (e.g. Arizona, California, Illinois, Mississippi, Missouri, New Mexico, Ohio, Pennsylvania
Tennessee, Texas, Utah, Virginia, possibly others)

We should probably have
- a really simple nux flow for them just to acknowledge that / remind them of that
- a super simplified just-one-tax-rate-for-you tax settings page

Core does something similar during its wizard:

screen shot 2017-05-24 at 2 25 13 pm

**FTW - since at this point in the flow we have the complete business address, we could actually use a service to get the rate and rate breakdown for their store and present it to them (as information only).  I see no need to allow them to edit it.**

See also #14439 and #14441

### LINKED ISSUE #14441: WooCommerce: Taxes: My store is in a state that collects sales tax based on where my orders ship to
Merchants in most states are required to collect sales tax based on where they ship product to.

We should probably have
- a really simple nux flow for them just to acknowledge that / remind them of that
- a super simplified tax settings page that mentions that, maybe gives the overall state rate but also notes that additional taxes vary by order shipping address and will be calculated automatically

See also #14439 and #14440

### PR DESCRIPTION
**Pull request by allendav**
Fixes #14439 
Fixes #14440 
Fixes #14441 
Fixes #13483

To test:

- After setting the store address during setup, navigate to store/settings/taxes/{site}
- Make sure the address is presented correctly
- Make sure you can edit the address. Rates should update when you edit the address.
- Make sure the tax information presented for the state and country are correct
- Make sure you can toggle taxes enabled on and off and save the change.
- Make sure you can toggle check boxes.
- Run `npm run test-client client/extensions/woocommerce/state`

### CODE REVIEW

File: client/extensions/woocommerce/lib/countries/US.js [hunk 1]
```diff
@@ -12,206 +12,257 @@ export default () => {
...
+					salesTaxBasis: 'destination',
 				},
 				{
 					code: 'DE',
 					name: translate( 'Delaware' ),
+					salesTaxBasis: 'none',
 				},
 				{
 					code: 'DC',
 					name: translate( 'District Of Columbia' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/lib/countries/US.js]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'District of Columbia' )` *ES Score: 22*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/lib/countries/US.js [hunk 2]
```diff
@@ -12,206 +21,257 @@ export default () => {
...
+					salesTaxBasis: DESTINATION_BASED_SALES_TAX,
 				},
 				{
 					code: 'DE',
 					name: translate( 'Delaware' ),
+					salesTaxBasis: NO_SALES_TAX,
 				},
 				{
 					code: 'DC',
 					name: translate( 'District Of Columbia' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/lib/countries/US.js]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'District of Columbia' )` *ES Score: 22*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+	onSave = ( event ) => {
+		const { site, translate } = this.props;
+
+		event.preventDefault();
+		this.setState( { isSaving: true } );
+
+		const onSuccess = () => {
+			this.setState( { isSaving: false, userBeganEditing: false } );
+			return successNotice( translate( 'Settings updated successfully.' ) );
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Theme settings updated successfully.' )` *ES Score: 14*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-tax-settings)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+			( <a href={ getLink( '/store/:site/', site ) }>{ translate( 'Settings' ) }</a> ),
+			( <span>{ translate( 'Taxes' ) }</span> ),
+		];
+
+		const saveButtonDisabled = this.state.isSaving || ! this.pageHasChanges();
+
+		return (
+			<Main className={ classNames( 'settings-taxes', className ) }>
+				<ActionHeader breadcrumbs={ breadcrumbs }>
+					<Button disabled={ saveButtonDisabled } onClick={ this.onSave } primary >
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: extra space after `primary`.

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	};
+}
+
+function mapDispatchToProps( dispatch ) {
+	return bindActionCreators(
+		{
+			fetchSettingsGeneral,
+			fetchTaxSettings,
+			updateTaxesEnabledSetting,
+			updateTaxSettings
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: Trailing comma.

File: client/extensions/woocommerce/app/settings/taxes/taxes-nexus.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		const { site } = this.props;
+
+		if ( site && site.ID ) {
+			this.props.fetchSettingsGeneral( site.ID );
+		}
+	}
+
+	onEditAddress = ( event ) => {
+		event.preventDefault();
+		// TODO - modal up
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/taxes-nexus.js]: See #15583. Looks like this whole component could be removed with mine just dropped in to the tax page.

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+		} ),
+	};
+
+	componentDidMount = () => {
+		const { address, loadedSettingsGeneral, loadedTaxRates, site } = this.props;
+
+		if ( site && site.ID ) {
+			if ( ! loadedSettingsGeneral ) {
+				this.props.fetchSettingsGeneral( site.ID );
+			} else if ( ! loadedTaxRates ) {
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: Is this just meant as a way to do one and then the other?

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+		return (
+			<div className="taxes__taxes-rates">
+				<ExtendedHeader
+					label={ translate( 'Tax rates' ) } >
+					<FormToggle
+						name="taxesEnabled"
+						onChange={ onEnabledChange }
+						checked={ taxesEnabled } >
+						<span>
+								{ toggleMessage }
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: Extra indents.

File: client/extensions/woocommerce/state/action-types.js [hunk 9]
```diff
@@ -1,3 +1,5 @@
+export const TAXRATES_REQUEST = 'TAXRATES_REQUEST';
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/action-types.js]: I feel like these should still be prefixed with `WOOCOMMERCE_`, as the reducers and actions are all under `woocommerce/state`.

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+			this.setState( { isSaving: false } );
+			return errorNotice( translate( 'There was a problem saving your changes. Please try again.' ) );
+		};
+
+		// TODO - chain these
+
+		this.props.updateTaxesEnabledSetting(
+			site.ID,
+			this.state.taxesEnabled,
+			null,
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: You can leave these `nulls` off since the defaults are null.

**Comment by justinshreve** [comment]
All tests pass except for one: 

`extensions woocommerce state sites product-categories reducer should not affect other state trees:`

`AssertionError: expected { general: {}, products: {}, tax: {} } to deeply equal { general: {}, products: {} }`

Though this looks like a fix needed for the test, not your code.

**Comment by justinshreve** [comment]
States don't seem to be loading at all on the address step.

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js (hunk 7 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: Yep - we can't request the tax rates until we have the store address from settings general

File: client/extensions/woocommerce/app/dashboard/pre-setup-view.js [hunk 11]
```diff
@@ -63,6 +63,13 @@ class PreSetupView extends Component {
 	onChange = ( event ) => {
 		const address = this.state.address;
 		address[ event.target.name ] = event.target.value;
+
+		// Did they change the country? Force an appropriate state default
+		if ( 'country' === name ) {
```
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/pre-setup-view.js]: Oops!

**Comment by kellychoffman** [comment]
Made some fixes and styled up some things :) Looking good! 

Just one thing - We got rid of the Site Title in the store location section.

File: client/extensions/woocommerce/state/action-types.js (hunk 9 above)
Comment (allendav) [inline client/extensions/woocommerce/state/action-types.js]: Agreed. Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js (hunk 8 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/settings/taxes/index.js (hunk 5 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/settings/taxes/index.js (hunk 4 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/settings/taxes/index.js (hunk 10 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: Agreed. Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/dashboard/pre-setup-view.js (hunk 11 above)
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/pre-setup-view.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15541/commits/17f50964f1f6156641032d53a6b96785ad2895a0

File: client/extensions/woocommerce/app/dashboard/pre-setup-view.js [hunk 12]
```diff
@@ -130,23 +137,25 @@ class PreSetupView extends Component {
...
+function mapStateToProps( state, ownProps ) {
 	let name = '';
-	if ( site ) {
-		name = getSiteTitle( state, site.ID );
+	let storeLocation = {};
+	let loading = true;
+
+	if ( ownProps.site ) {
+		name = getSiteTitle( state, ownProps.site.ID );
+		storeLocation = getStoreLocation( state, ownProps.site.ID );
```
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/pre-setup-view.js]: Agreed! Swapped out in https://github.com/Automattic/wp-calypso/pull/15541/commits/5552a8792688f39b535c77718f94bdaf20514ec9

File: client/extensions/woocommerce/app/settings/taxes/taxes-nexus.js (hunk 6 above)
Comment (allendav) [inline client/extensions/woocommerce/app/settings/taxes/taxes-nexus.js]: Agreed! Swapped out in https://github.com/Automattic/wp-calypso/pull/15541/commits/5552a8792688f39b535c77718f94bdaf20514ec9

**Pull request merged**